crates/sui-types/src/nitro_attestation.rs

```diff
@@ -0,0 +1,607 @@
+    {
+        use serde::de::{Error, SeqAccess, Visitor};
+        use std::fmt;
+
+        struct CoseSign1Visitor;
+
+        impl<'de> Visitor<'de> for CoseSign1Visitor {
+            type Value = CoseSign1;
+
+            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                f.write_str("a possibly tagged CoseSign1 structure")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<CoseSign1, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                fn extract_bytes(value: Value) -> Option<Vec<u8>> {
+                    match value {
+                        Value::Bytes(bytes) => Some(bytes),
+                        _ => None,
+                    }
+                }
+
+                // Get protected header bytes
+                let protected = match seq.next_element::<Value>()? {
+                    Some(v) => extract_bytes(v)
+                        .ok_or_else(|| A::Error::custom("protected header must be bytes"))?,
+                    None => return Err(A::Error::missing_field("protected")),
+                };
+
+                let unprotected = match seq.next_element()? {
+                    Some(v) => v,
+                    None => return Err(A::Error::missing_field("unprotected")),
+                };
+                // Get payload bytes
+                let payload = match seq.next_element::<Value>()? {
+                    Some(v) => {
+                        extract_bytes(v).ok_or_else(|| A::Error::custom("payload must be bytes"))?
+                    }
+                    None => return Err(A::Error::missing_field("payload")),
+                };
+
+                // Get signature bytes
+                let signature = match seq.next_element::<Value>()? {
+                    Some(v) => extract_bytes(v)
+                        .ok_or_else(|| A::Error::custom("signature must be bytes"))?,
+                    None => return Err(A::Error::missing_field("signature")),
+                };
+
+                Ok(CoseSign1 {
+                    protected,
+                    unprotected,
+                    payload,
+                    signature,
+                })
+            }
+
+            fn visit_newtype_struct<D>(self, deserializer: D) -> Result<CoseSign1, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                // This is the tagged version: we ignore the tag part, and just go into it
+                deserializer.deserialize_seq(CoseSign1Visitor)
+            }
+        }
+
+        deserializer.deserialize_any(CoseSign1Visitor)
+    }
+}
+
+impl CoseSign1 {
+    /// Parse CBOR bytes into struct. Adapted from <https://github.com/awslabs/aws-nitro-enclaves-cose/blob/main/src/sign.rs>
+    pub fn parse_and_validate(bytes: &[u8]) -> Result<Self, NitroAttestationVerifyError> {
+        let tagged_value: ciborium::value::Value = ciborium::de::from_reader(bytes)
```

Comment:
general question - is ciborium decoding linear in the input? (if it's quadratic or worse we might want more checks)

Reply:
yes its linear- the parsing basically iterate the entire document_map per variable name to parse each value. we charge per bytes of the total attestation bytes. 

however for the pcrs, it is quadratic wrt to the number of pcrs here. but there should only be 6 valid indices so its bounded.